### PR TITLE
fix false positive kubectl plugin unit tests

### PR DESCRIPTION
**Pull request by atiratree**
#### What type of PR is this?

/kind cleanup
/kind failing-test
^ kind test-fails-at-failing would be more appropriate

#### What this PR does / why we need it:

I have been playing with TestKubectlCommandHandlesPlugins and TestKubectlSubcommandShadowPlugin tests and noticed that they actually do not run properly. Our goal is to test different commands to test conflicts with plugins, but the test mostly just executes without args (just `kubectl`), depending on how you run the tests. This in turn just calls help and exists with 0 and no conflicts are tested. This PR tries to remedy that to prevent future regressions in the plugin logic.

- test.args should be passed instead of the os.Args of the test framework to prevent simple invocation of kubectl without args that could manifest in false positive test runs
- plugin execution should have a different test path
- tests should invoke functioning kubectl commands instead of the mock ones to ensure the correct subcommand is executed without a failure

@ardaguclu @soltysh 

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go [hunk 1]
```diff
@@ -73,21 +74,21 @@ func TestKubectlSubcommandShadowPlugin(t *testing.T) {
...
+			args:             []string{"kubectl", "create", "job", "foo", "--image=busybox", "--dry-run=client", "--namespace", "test-namespace"},
 			expectPlugin:     "",
 			expectPluginArgs: []string{},
 		},
 		// rest of the tests are copied from TestKubectlCommandHandlesPlugins function,
 		// just to retest them also when feature is enabled.
 		{
 			name:             "test that normal commands are able to be executed, when no plugin overshadows them",
-			args:             []string{"kubectl", "get", "foo"},
+			args:             []string{"kubectl", "config", "get-clusters"},
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go:91]: What is the reason of this change?

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go [hunk 2]
```diff
@@ -99,19 +100,19 @@ func TestKubectlSubcommandShadowPlugin(t *testing.T) {
 		},
 		{
 			name: "test that a plugin does not execute over an existing command by the same name",
-			args: []string{"kubectl", "version"},
+			args: []string{"kubectl", "version", "--client=true"},
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go:103]: to assure that we are executing builtin because only builtin `version` command has `--client` flag?

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go [hunk 3]
```diff
@@ -132,19 +133,32 @@ func TestKubectlSubcommandShadowPlugin(t *testing.T) {
...
+					root.SetArgs(test.args[1:])
+					// Important note! Incorrect command or command failing validation might just call os.Exit(1) which would interrupt execution of the test
+					if err := root.Execute(); err != nil {
+						t.Fatalf("unexpected error: %v", err)
+					}
 				}
 
-				if pluginsHandler.err != nil && pluginsHandler.err.Error() != test.expectError {
-					t.Fatalf("unexpected error: expected %q to occur, but got %q", test.expectError, pluginsHandler.err)
+				if pluginsHandler.lookupErr != nil && pluginsHandler.lookupErr.Error() != test.expectLookupError {
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go:153]: I understand the reason of renaming of `err` to `lookupErr` but it is hard to follow the real changes, I think it would be better to do it in separate commit.

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go [hunk 4]
```diff
@@ -236,19 +250,32 @@ func TestKubectlCommandHandlesPlugins(t *testing.T) {
...
+				// args must be set, otherwise Execute will use os.Args (args used for starting the test) and test.args would not be passed
+				// to the command which might invoke only "kubectl" without any additional args and give false positives
+				root.SetArgs(test.args[1:])
+				// Important note! Incorrect command or command failing validation might just call os.Exit(1) which would interrupt execution of the test
+				if err := root.Execute(); err != nil {
+					t.Fatalf("unexpected error: %v", err)
+				}
+			}
+
+			if pluginsHandler.lookupErr != nil && pluginsHandler.lookupErr.Error() != test.expectLookupError {
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go:270]: Where is the test case of `LookupError` is expected but not fired?

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go [hunk 5]
```diff
@@ -236,19 +250,32 @@ func TestKubectlCommandHandlesPlugins(t *testing.T) {
...
 			}
 
-			if pluginsHandler.err != nil && pluginsHandler.err.Error() != test.expectError {
-				t.Fatalf("unexpected error: expected %q to occur, but got %q", test.expectError, pluginsHandler.err)
+			if pluginsHandler.lookedup && !pluginsHandler.executed && len(test.expectLookupError) == 0 {
+				t.Fatalf("expected plugin execution, but did not occur")
 			}
 
-			if pluginsHandler.executedPlugin != test.expectPlugin {
+			if pluginsHandler.executedPlugin != test.expectPlugin || (pluginsHandler.executed && test.expectPlugin == "") {
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go:278]: `(pluginsHandler.executed && test.expectPlugin == "")` could be in separate condition with different message

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go [hunk 6]
```diff
@@ -236,19 +250,32 @@ func TestKubectlCommandHandlesPlugins(t *testing.T) {
...
 			pluginsHandler := &testPluginHandler{
 				pluginsDirectory: "plugin/testdata",
+				validPrefixes:    plugin.ValidPluginFilenamePrefixes,
 			}
 			ioStreams, _, _, _ := genericiooptions.NewTestIOStreams()
-
 			root := NewDefaultKubectlCommandWithArgs(KubectlOptions{PluginHandler: pluginsHandler, Arguments: test.args, IOStreams: ioStreams})
-			if err := root.Execute(); err != nil {
-				t.Fatalf("unexpected error: %v", err)
+			if pluginsHandler.lookedup || pluginsHandler.executed {
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go:257]: We haven't run `root.Execute` yet, how `lookedup` or `executed` can be true?

**Review by ardaguclu — COMMENTED** [review]
Thanks @atiratree. 
Could you please point out how we can reproduce this false positive unit tests by adding which flag into which command?

**Comment by atiratree** [comment]
>  Could you please point out how we can reproduce this false positive unit tests by adding which flag into which command?

It cannot be reproduced by adding a flag. This just ensures that a regression cannot be introduced into the command/plugin resolution code. We want to test that a correct command is executed and no error is returned.

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go (hunk 1 above)
Comment (atiratree) [reply staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go:91]: these are changed to commands that require no cluster connection, so they can be tested without failing

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go (hunk 2 above)
Comment (atiratree) [reply staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go:103]: same here, in order not to need a cluster connection

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go (hunk 3 above)
Comment (atiratree) [reply staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go]: ack,  moved to a separate commit

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go (hunk 4 above)
Comment (atiratree) [reply staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go]: +1, added to the condition

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go (hunk 5 above)
Comment (atiratree) [reply staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go]: ack, moved to a separate condition

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go (hunk 6 above)
Comment (atiratree) [reply staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go:262]: because DefaultPluginHandler is implemented by exec call as commented bellow, so `root.Execute` will never be called in the end

File: staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go [hunk 7]
```diff
@@ -132,22 +133,40 @@ func TestKubectlSubcommandShadowPlugin(t *testing.T) {
...
 				pluginsHandler := &testPluginHandler{
 					pluginsDirectory: "plugin/testdata",
+					validPrefixes:    plugin.ValidPluginFilenamePrefixes,
 				}
 				ioStreams, _, _, _ := genericiooptions.NewTestIOStreams()
-
 				root := NewDefaultKubectlCommandWithArgs(KubectlOptions{PluginHandler: pluginsHandler, Arguments: test.args, IOStreams: ioStreams})
-				if err := root.Execute(); err != nil {
-					t.Fatalf("unexpected error: %v", err)
+				if pluginsHandler.lookedup || pluginsHandler.executed {
```
Comment (ardaguclu) [inline staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go:140]: Can we simplify this to?;

```go
if !pluginsHandler.lookedup && !pluginsHandler.executed {
        root.SetArgs(test.args[1:])
        if err := root.Execute(); err != nil {
              t.Fatalf("unexpected error: %v", err)
        }
}
```

Comment (atiratree) [reply staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go:140]: I can change it if you prefer, but I think it better shows the intent of the test if we make the if/else more verbose. Especially when there are multiple comments for each section, so I would opt for better readability in the test code.

Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go:140]: I'm fine adding comments, especially
```
// args must be set, otherwise Execute will use os.Args (args used for starting the test) and test.args would not be passed
// to the command which might invoke only "kubectl" without any additional args and give false positives
```
seems pretty useful to explain why we are passing `test.args[1:]` explicitly. 

But considering that the other comments can also be extracted by the person who work on unit tests by debugging the code and I'd not prefer sacrificing conventions(no code, only comment between if - else).

Comment (ardaguclu) [reply staging/src/k8s.io/kubectl/pkg/cmd/cmd_test.go]: Sounds good, thanks

**Comment by ardaguclu** [comment]
Thanks @atiratree 
/lgtm
/approve
